internal/provider/resource_cert_request_test.go

```diff
@@ -49,86 +48,76 @@ func TestCertRequest(t *testing.T) {
-                        value = "${tls_cert_request.test1.cert_request_pem}"
-                    }
                 `, testPrivateKeyPEM),
-				Check: func(s *terraform.State) error {
-					gotUntyped := s.RootModule().Outputs["key_pem_1"].Value
-
-					got, ok := gotUntyped.(string)
-					if !ok {
-						return fmt.Errorf("output for \"key_pem_1\" is not a string")
-					}
-
-					if !strings.HasPrefix(got, "-----BEGIN CERTIFICATE REQUEST----") {
-						return fmt.Errorf("key is missing CSR PEM preamble")
-					}
-					block, _ := pem.Decode([]byte(got))
-					csr, err := x509.ParseCertificateRequest(block.Bytes)
-					if err != nil {
-						return fmt.Errorf("error parsing CSR: %s", err)
-					}
-					if expected, got := "2", csr.Subject.SerialNumber; got != expected {
-						return fmt.Errorf("incorrect subject serial number: expected %v, got %v", expected, got)
-					}
-					if expected, got := "example.com", csr.Subject.CommonName; got != expected {
-						return fmt.Errorf("incorrect subject common name: expected %v, got %v", expected, got)
-					}
-					if expected, got := "Example, Inc", csr.Subject.Organization[0]; got != expected {
-						return fmt.Errorf("incorrect subject organization: expected %v, got %v", expected, got)
-					}
-					if expected, got := "Department of Terraform Testing", csr.Subject.OrganizationalUnit[0]; got != expected {
-						return fmt.Errorf("incorrect subject organizational unit: expected %v, got %v", expected, got)
-					}
-					if expected, got := "5879 Cotton Link", csr.Subject.StreetAddress[0]; got != expected {
-						return fmt.Errorf("incorrect subject street address: expected %v, got %v", expected, got)
-					}
-					if expected, got := "Pirate Harbor", csr.Subject.Locality[0]; got != expected {
-						return fmt.Errorf("incorrect subject locality: expected %v, got %v", expected, got)
-					}
-					if expected, got := "CA", csr.Subject.Province[0]; got != expected {
-						return fmt.Errorf("incorrect subject province: expected %v, got %v", expected, got)
-					}
-					if expected, got := "US", csr.Subject.Country[0]; got != expected {
-						return fmt.Errorf("incorrect subject country: expected %v, got %v", expected, got)
-					}
-					if expected, got := "95559-1227", csr.Subject.PostalCode[0]; got != expected {
-						return fmt.Errorf("incorrect subject postal code: expected %v, got %v", expected, got)
-					}
+				Check: r.ComposeAggregateTestCheckFunc(
+					r.TestMatchResourceAttr("tls_cert_request.test1", "cert_request_pem", regexp.MustCompile(`^-----BEGIN CERTIFICATE REQUEST----`)),
+					r.TestCheckResourceAttrWith("tls_cert_request.test1", "cert_request_pem", func(value string) error {
+						block, _ := pem.Decode([]byte(value))
+						csr, err := x509.ParseCertificateRequest(block.Bytes)
+						if err != nil {
+							return fmt.Errorf("error parsing CSR: %s", err)
+						}
+						if expected, got := "2", csr.Subject.SerialNumber; got != expected {
```

Comment:
Nit: Since there are repeating checks for `*x509.CertificateRequest` contents, would it be easier to introduce a helper function in this provider codebase? e.g.

```go
func testCheckPEMCertificateRequest(expected *x509.CertificateRequest) resource.TestCheckResourceAttrWithFunc {
  return func(value string) error {
    // decode and parse
    // equality logic (bummer there's not an Equal() method on it)
  }
}
```

Then the test can remain declarative:

```go
r.TestCheckResourceAttrWith("tls_cert_request.test1", "cert_request_pem", testCheckPEMCertificateRequest(&x509.CertificateRequest{
  Subject: x509.Subject{
    CommonName: "example.com",
    SerialNumber: 2,
    // ... etc
  },
  // ... etc
}),
```

Reply:
(or the test helper can wrap away the `TestCheckResourceAttrWith()` completely, e.g.)

```go
func testCheckPEMCertificateRequest(name string, key string, expected *x509.CertificateRequest) resource.TestCheckFunc {
  return func(s *terraform.State) error {
    return resource.TestCheckResourceAttrWith(name, key, func(value string) error {
      // decode and parse
      // equality logic (bummer there's not an Equal() method on it)
    })(s)
  }
}
```

Reply:
Will give this a go.